### ISSUE TITLE
Sleeping task upstream from many tasks causes hang

### ISSUE THREAD

**Issue by tekumara** [PrefectHQ/prefect#6961]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

Tasks downstream of an upstream task that sleeps are never executed and the flow just hangs.

If the sleep is shorter, or there are fewer tasks, the flow completes.

### Reproduction

```python
from time import sleep
from typing import List

from prefect import flow, get_run_logger, task


@task
def setup() -> None:
    logger = get_run_logger()
    logger.info("setup start")
    # sleep(1) won't hang
    sleep(3)
    logger.info("setup end")


@task
def fetch_batches() -> List[str]:
    # using range(30) here won't hang
    return [f"batch {i}" for i in range(50)]


@task
def count_rows(batch: str) -> int:
    logger = get_run_logger()
    logger.info(f"{batch}")
    return 1


@flow
def map_flow() -> None:
    batches = fetch_batches.submit()
    count_rows.map(batches, wait_for=[setup.submit()])  # type: ignore


if __name__ == "__main__":
    map_flow()
```

### Error

```
 python -m flows.map_flow
19:25:40.025 | INFO    | prefect.engine - Created flow run 'lilac-marmoset' for flow 'map-flow'
19:25:40.241 | INFO    | Flow run 'lilac-marmoset' - Created task run 'fetch_batches-deeb0ec9-0' for task 'fetch_batches'
19:25:40.242 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'fetch_batches-deeb0ec9-0' for execution.
19:25:40.260 | INFO    | Flow run 'lilac-marmoset' - Created task run 'setup-8332e509-0' for task 'setup'
19:25:40.260 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'setup-8332e509-0' for execution.
19:25:40.317 | INFO    | Task run 'setup-8332e509-0' - setup start
19:25:40.326 | INFO    | Task run 'fetch_batches-deeb0ec9-0' - Finished in state Completed()
19:25:40.487 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-34' for task 'count_rows'
19:25:40.487 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-34' for execution.
19:25:40.494 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-17' for task 'count_rows'
19:25:40.494 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-17' for execution.
19:25:40.502 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-20' for task 'count_rows'
19:25:40.503 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-20' for execution.
19:25:40.511 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-10' for task 'count_rows'
19:25:40.512 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-10' for execution.
19:25:40.521 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-3' for task 'count_rows'
19:25:40.521 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-3' for execution.
19:25:40.532 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-11' for task 'count_rows'
19:25:40.532 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-11' for execution.
19:25:40.540 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-35' for task 'count_rows'
19:25:40.540 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-35' for execution.
19:25:40.555 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-44' for task 'count_rows'
19:25:40.555 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-44' for execution.
19:25:40.569 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-16' for task 'count_rows'
19:25:40.569 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-16' for execution.
19:25:40.585 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-39' for task 'count_rows'
19:25:40.585 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-39' for execution.
19:25:40.597 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-7' for task 'count_rows'
19:25:40.597 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-7' for execution.
19:25:40.611 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-19' for task 'count_rows'
19:25:40.611 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-19' for execution.
19:25:40.623 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-30' for task 'count_rows'
19:25:40.623 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-30' for execution.
19:25:40.637 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-29' for task 'count_rows'
19:25:40.638 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-29' for execution.
19:25:40.656 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-1' for task 'count_rows'
19:25:40.656 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-1' for execution.
19:25:40.669 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-14' for task 'count_rows'
19:25:40.669 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-14' for execution.
19:25:40.682 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-9' for task 'count_rows'
19:25:40.682 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-9' for execution.
19:25:40.703 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-23' for task 'count_rows'
19:25:40.703 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-23' for execution.
19:25:40.718 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-31' for task 'count_rows'
19:25:40.718 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-31' for execution.
19:25:40.730 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-24' for task 'count_rows'
19:25:40.730 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-24' for execution.
19:25:40.751 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-41' for task 'count_rows'
19:25:40.751 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-41' for execution.
19:25:40.810 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-36' for task 'count_rows'
19:25:40.810 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-36' for execution.
19:25:40.821 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-4' for task 'count_rows'
19:25:40.821 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-4' for execution.
19:25:40.835 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-43' for task 'count_rows'
19:25:40.835 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-43' for execution.
19:25:40.917 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-12' for task 'count_rows'
19:25:40.918 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-12' for execution.
19:25:40.935 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-21' for task 'count_rows'
19:25:40.936 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-21' for execution.
19:25:40.956 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-49' for task 'count_rows'
19:25:40.956 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-49' for execution.
19:25:41.023 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-15' for task 'count_rows'
19:25:41.023 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-15' for execution.
19:25:41.038 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-46' for task 'count_rows'
19:25:41.039 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-46' for execution.
19:25:41.054 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-45' for task 'count_rows'
19:25:41.054 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-45' for execution.
19:25:41.130 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-32' for task 'count_rows'
19:25:41.130 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-32' for execution.
19:25:41.145 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-40' for task 'count_rows'
19:25:41.146 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-40' for execution.
19:25:41.238 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-13' for task 'count_rows'
19:25:41.238 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-13' for execution.
19:25:41.262 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-33' for task 'count_rows'
19:25:41.263 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-33' for execution.
19:25:41.344 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-2' for task 'count_rows'
19:25:41.344 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-2' for execution.
19:25:41.355 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-25' for task 'count_rows'
19:25:41.356 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-25' for execution.
19:25:41.438 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-0' for task 'count_rows'
19:25:41.438 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-0' for execution.
19:25:41.454 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-5' for task 'count_rows'
19:25:41.455 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-5' for execution.
19:25:41.540 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-6' for task 'count_rows'
19:25:41.540 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-6' for execution.
19:25:41.612 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-47' for task 'count_rows'
19:25:41.612 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-47' for execution.
19:25:41.647 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-28' for task 'count_rows'
19:25:41.647 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-28' for execution.
19:25:41.670 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-37' for task 'count_rows'
19:25:41.670 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-37' for execution.
19:25:41.751 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-8' for task 'count_rows'
19:25:41.751 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-8' for execution.
19:25:41.777 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-42' for task 'count_rows'
19:25:41.777 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-42' for execution.
19:25:41.851 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-27' for task 'count_rows'
19:25:41.851 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-27' for execution.
19:25:41.956 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-18' for task 'count_rows'
19:25:41.956 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-18' for execution.
19:25:42.061 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-22' for task 'count_rows'
19:25:42.061 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-22' for execution.
19:25:42.162 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-38' for task 'count_rows'
19:25:42.162 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-38' for execution.
19:25:42.269 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-26' for task 'count_rows'
19:25:42.270 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-26' for execution.
19:25:42.381 | INFO    | Flow run 'lilac-marmoset' - Created task run 'count_rows-41aff3d6-48' for task 'count_rows'
19:25:42.381 | INFO    | Flow run 'lilac-marmoset' - Submitted task run 'count_rows-41aff3d6-48' for execution.
19:25:43.322 | INFO    | Task run 'setup-8332e509-0' - setup end
```

### Versions

```
Version:             2.4.2
API version:         0.8.0
Python version:      3.9.13
Git commit:          65807e84
Built:               Fri, Sep 23, 2022 10:43 AM
OS/Arch:             darwin/arm64
Profile:             default
Server type:         ephemeral
Server:
  Database:          sqlite
  SQLite version:    3.37.0
```

**Comment by madkinsz** [comment]
Thanks for the MRE! I believe this may be addressed by #6948 — can you try to reproduce with that branch?

**Comment by tekumara** [comment]
It's still happening when I use commit https://github.com/PrefectHQ/prefect/pull/6948/commits/464087238fe1df7c7f698eb3c85c1431eaeeaa4f 😿

**Comment by zhen0** [comment]
Ran the MRE flow locally on 2.8.7 and it completes.

**Comment by madkinsz** [comment]
Addressed by #8887 and #8702
